sqlglot/expressions.py

```diff
@@ -219,10 +220,21 @@ def type(self, dtype: t.Optional[DataType | DataType.Type | str]) -> None:
             dtype = DataType.build(dtype)
         self._type = dtype  # type: ignore
 
+    @property
+    def metadata(self) -> t.Optional[t.Dict[str, t.Any]]:
+        return self._meta
+
+    def set_metadata(self, meta_key, meta_value):
+        if not self._meta:
+            self._meta = {}
+
+        self._meta[meta_key] = meta_value
+
     def __deepcopy__(self, memo):
         copy = self.__class__(**deepcopy(self.args))
         copy.comments = self.comments
         copy.type = self.type
+        copy._meta = self.metadata
```

Comment:
this doesn't perform a deep copy of metadata, i don't think this is correct

Reply:
now that i look at it comments is wrong too

if self._meta is not None:
    copy.meta = deecopy(self.meta) 

do something similar for copy.comments

Reply:
Yea that slipped, `type` is wrong too I think. We need to do `.copy()` if it's not `None`.